docs/DesignDocs/Protected_Code_Loader_Proposal.md

```diff
@@ -0,0 +1,161 @@
+Open Enclave Protected Code Loader
+====
+
+Enabling Enclave Code Confidentiality in Open Enclave.
+
+# Motivation
+Current Open Enclave SDK provides code integrity and data confidentiality at
+run-time but not enclave binary confidentiality on a disk. The enclave binary
+is in plaintext and can be reverse engineered to reveal code logic and
+secret data embedded in the enclave image. 
+
+The Open Enclave Protected Code Loader shall enable providing confidentiality 
+and integrity to the IP sections in enclave image, i.e. , the ELF file.
+
+# User Experience
+## Enclave Build-time:
+### Linking pcl lib to the enclave 
+On building the enclave, a static lib "liboepcl.a" should be linked to the 
+enclave image with the compilation option "-Wl, whole-archive"
+### Encrypt the enclave image
+An encryption tool is provided as the pre-step to enclave signing process:
+
+```bash
+oeencrypt -i my_enclave -o my_enclave.enc -k keyfile
+```
+
+## Enclave Load-time:
+On Enclave Loading, if the image is encrypted, it need to be decrypted 
+inside enclave loader, this step is done by OE SDK runtime and invisible 
+to user. 
+
+# Specification
+
+## Protected Code Loader Software Work Flow in Open Enclave
+### Build time encryption 
+    A seperate tool "oeencrypt" is provided at enclave build time to encrypt 
+	the enclave image right before enclave signing process.  
+### Load time decryption
+    On enclave initliazation operation(EINIT operation), right before relocation, 
+    the encrypted enclave image need to be decrypted and than perform the relocation
+    operation.
+
+## Encryption Algorithm in Protected Code Loader
+Using openssl AES-256-GCM as the encryption/decryption algorithm.
+
+## ELF sections Left plaintext
```

Comment:
Putting a comment I had from the previous SIG-Arch meeting in writing: Have we considered if this design can be simplified if the OE SDK loader simply supported dynamic binding on load? (i.e. an enclave app can consist of multiple shared objects and not only a statically linked binary)

It seems like that could simplify the separation of encrypted/unencrypted boundaries by simply marking individual binaries as encrypted or not, instead of having to redeclare custom sections of the ELF binary.